### PR TITLE
510 UI for oauth2

### LINKED ISSUE #510: ui elements for oauth2 authorisation workflows
We need to add proper ui workflows for oauth2 authorisation for users to interact with, that can also handle unauthenticated accounts at the same time.

### LINKED ISSUE #518: WebUI - enter key submission of login fields
When logging in, pressing enter doesn't submit the username/TOTP/password fields, which requires the user to keep changing input methods.

It should be solve-able with something along these lines: http://stackoverflow.com/questions/18160342/ddg#18160418

```javascript
$('#username').keypress(function (e) {
 var key = e.which;
 if(key == 13)  // the enter key code
  {
    $('input[name = usernamebutton]').click();
    return false;  
  }
});
```

This uses the existing jquery code already included.

### PR DESCRIPTION
**Pull request by Firstyear**
Fixes #510
Fixes #518
- this adds the needed ui elements for oauth2 workflows. There are surely still rough edges, but it works and handles the majority of current edge cases. 

- [ x ] cargo fmt has been run
- [ - ] cargo clippy has been run
- [ x ] cargo test has been run and passes
- [ - ] book chapter included (if relevant)
- [ - ] design document included (if relevant)

### CODE REVIEW

**Comment by Firstyear** [comment]
I probably need to clean up some compiler warnings before merging this still.

**Comment by Firstyear** [comment]
Okay, these are fixed, it's ready for review

**Review by QnnOkabayashi — COMMENTED** [review]
To be honest I didn't understand of lot of this because it's mostly library calls. Is there an easy way I can few what these changes will do to the web UI? Or see screenshots?

File: kanidmd/src/lib/core/https/mod.rs [hunk 1]
```diff
@@ -1222,9 +1226,13 @@ pub fn create_https_server(
...
     let mut oauth2_process = tserver.at("/oauth2");
-    oauth2_process.at("/authorise").get(oauth2_authorise_get);
+    oauth2_process
+        .at("/authorise")
+        .post(oauth2_authorise_post)
+        .get(oauth2_authorise_get);
     oauth2_process
         .at("/authorise/permit")
+        .post(oauth2_authorise_permit_post)
         .get(oauth2_authorise_permit_get);
```
[the remark this answers is not in the document]
Comment (Firstyear) [reply kanidmd/src/lib/core/https/mod.rs:1236]: For the same url but with different http actions. get and post have different semantics on how arguments are passed to these end points.

File: kanidmd_web_ui/pkg/package.json [hunk 2]
```diff
@@ -4,7 +4,7 @@
     "William Brown <william@blackhats.net.au>"
   ],
   "description": "Kanidm Server Web User Interface",
-  "version": "1.1.0-alpha.4",
```
[the remark this answers is not in the document]
Comment (Firstyear) [reply kanidmd_web_ui/pkg/package.json:7]: Not at all! There is still token introspection and scopes to be added.

File: kanidmd_web_ui/pkg/style.css [hunk 3]
```diff
@@ -0,0 +1,141 @@
+.html-body {
```
[the remark this answers is not in the document]
Comment (Firstyear) [reply kanidmd_web_ui/pkg/style.css:1]: If you run the server locally and navigate to it in your browser you'll see it :)

File: kanidmd_web_ui/src/manager.rs [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+                if self.is_ready {
+                    html! {<Router<Route> render=Router::render(switch) /> }
+                } else {
+                    html! { <body></body> }
+                }
+            }
+        </>
+        }
+    }
+}
```
[the remark this answers is not in the document]
Comment (Firstyear) [reply kanidmd_web_ui/src/manager.rs:121]: Yeah, they are seriously magic. I'm too stupid for procmacros but they are wildly powerful.

File: kanidmd_web_ui/src/oauth2.rs [hunk 5]
```diff
@@ -0,0 +1,360 @@
...
+    }
+
+    fn update(&mut self, msg: Self::Message) -> ShouldRender {
+        ConsoleService::log("oauth2::update");
+
+        match msg {
+            Oauth2Msg::LoginProceed => {
+                models::push_return_location(models::Location::Oauth2);
+                yew_router::push_route(Route::Login);
+                // Don't need to redraw as we are yolo-ing out.
```
[the remark this answers is not in the document]
Comment (Firstyear) [reply kanidmd_web_ui/src/oauth2.rs:233]: Yep, once we leave the update function, because we pushed a new route, the top level manager we redirect us out into the login view.

File: kanidmd_web_ui/pkg/bundle.js [hunk 6]
```diff
@@ -485,14 +504,52 @@
...
+        imports.wbg.__wbg_warn_b39e749f1dc02058 = function(arg0) {
+            console.warn(getObject(arg0));
+        };
+        imports.wbg.__wbg_instanceof_HtmlElement_835072e813858ac0 = function(arg0) {
+            var ret = getObject(arg0) instanceof HTMLElement;
+            return ret;
+        };
+        imports.wbg.__wbg_focus_2fac919cca20d33b = function() { return handleError(function (arg0) {
+            getObject(arg0).focus();
+        }, arguments) };
```
[the remark this answers is not in the document]
Comment (yaleman) [reply kanidmd_web_ui/pkg/bundle.js:516]: This is all automated code generated by wasm, basically transpiling rust into ... something the browser can use. Thankfully, @Firstyear didn't have to write all this :)

File: kanidmd_web_ui/src/manager.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (yaleman) [reply kanidmd_web_ui/src/manager.rs:121]: Ooh, this sure is pretty compared to... bucket-of-strings. 😻

**Comment by yaleman** [comment]
I'm dumb, how do I test this? 😄 I get a sessionid error trying to log into the main interface and the URIs I can find in the source aren't much help 😊

**Review by QnnOkabayashi — APPROVED** [review]

**Comment by Firstyear** [comment]
> I'm dumb, how do I test this? 😄 I get a sessionid error trying to log into the main interface and the URIs I can find in the source aren't much help 😊

Without an oauth2 client application you may not be able to test that part of the flow. You may also need to clear your local session storage (cookies etc). Do you know what part you are getting the sessionid error on? We could have some better handling for clearing this automatically in error conditions.

**Comment by yaleman** [comment]
> > I'm dumb, how do I test this? 😄 I get a sessionid error trying to log into the main interface and the URIs I can find in the source aren't much help 😊
> 
> Without an oauth2 client application you may not be able to test that part of the flow. You may also need to clear your local session storage (cookies etc). Do you know what part you are getting the sessionid error on? We could have some better handling for clearing this automatically in error conditions.

I'll log it as another issue - as soon as I type in my username and click "begin" I get this:
Screen Shot 2021-07-24 at 17 07 56

I tried using https://oauthdebugger.com and https://oidcdebugger.com (tried all the various options on the page, sending a request to `https://myhostname.internal/oauth2/authorise`, I'm guessing they doesn't support the PKCE oauth2 flow you're targeting. 😞  as I got this error in the logs:

    [kanidm::core::https::oauth2] failed with reason: missing field `code_challenge`

I can't find a test app online for this flow so I'll have to pass on reviewing it further 😄

**Comment by Firstyear** [comment]
@yaleman I think that's an issue with cookies, I've seen it before. I think there needs to be some better handling around partial login session cleaning. I think it should be opened as an issue so I can investigate it further. 

Anyway, I think quinn's given it the okay, you okay for me to merge too?

**Review by yaleman — APPROVED** [review]
Yeet the code in, captain!
